pkg/controller/syncclusterdeployment/syncclusterdeployment_controller.go

```diff
@@ -0,0 +1,445 @@
+// This function is not meant to be invoked concurrently with the same key.
+func (c *Controller) syncClusterDeployment(key string) error {
+	startTime := time.Now()
+	c.logger.WithField("key", key).Debug("syncing machineset")
+	defer func() {
+		c.logger.WithFields(log.Fields{
+			"key":      key,
+			"duration": time.Now().Sub(startTime),
+		}).Debug("finished syncing machineset")
+	}()
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	cluster, err := c.clusterLister.ClusterDeployments(namespace).Get(name)
+	if err != nil {
+		c.logger.Errorf("Error fetching cluster %v/%v", namespace, name)
+		return err
+	}
+
+	err = c.syncMachineSets(cluster)
+
+	return err
+}
+
+// makes sure the remote cluster's MachineSets match the provided ClusterDeployment.Spec.MachineSets[]
+func (c *Controller) syncMachineSets(cluster *cov1.ClusterDeployment) error {
+	if !cluster.Status.ClusterAPIInstalled {
+		c.logger.Debugf(errorMsgClusterAPINotInstalled)
+		return nil
+	}
+
+	machineSetsToDelete := []*clusterapiv1.MachineSet{}
+	machineSetsToCreate := []*clusterapiv1.MachineSet{}
+	machineSetsToUpdate := []*clusterapiv1.MachineSet{}
+
+	remoteClusterAPIClient, err := c.BuildRemoteClient(cluster)
+	if err != nil {
+		return err
+	}
+	remoteMachineSets, err := remoteClusterAPIClient.ClusterV1alpha1().MachineSets(remoteClusterAPINamespace).List(metav1.ListOptions{})
+
+	coVersionNamespace, coVersionName := cluster.Spec.ClusterVersionRef.Namespace, cluster.Spec.ClusterVersionRef.Name
+	coClusterVersion, err := c.client.ClusteroperatorV1alpha1().ClusterVersions(coVersionNamespace).Get(coVersionName, metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("cannot retrieve cluster version %s/%s: %v", coVersionNamespace, coVersionName, err)
+	}
+
+	clusterDeploymentComputeMachineSets := computeClusterAPIMachineSetsFromClusterDeployment(cluster, coClusterVersion)
+
+	// find items that need updating/creating
+	for _, ms := range clusterDeploymentComputeMachineSets {
+		found := false
+		for _, rMS := range remoteMachineSets.Items {
+			if ms.Name == rMS.Name {
+				found = true
+				// TODO what other fields do we want to check for?
+				if *rMS.Spec.Replicas != *ms.Spec.Replicas {
+					machineSetsToUpdate = append(machineSetsToUpdate, ms)
+				}
+				break
+			}
+		}
+
+		if !found {
+			machineSetsToCreate = append(machineSetsToCreate, ms)
+		}
+
+	}
+
+	// find items that need deleting
+	for _, rMS := range remoteMachineSets.Items {
+		found := false
+		for _, ms := range clusterDeploymentComputeMachineSets {
+			if rMS.Name == ms.Name {
+				found = true
+				break
+			}
+		}
+		if !found {
+			machineSetsToDelete = append(machineSetsToDelete, &rMS)
+		}
+	}
+
+	c.logger.Debugf("Creating machineset(s) %+v", machineSetsToCreate)
+	c.logger.Debugf("Updating machineset(s) %+v", machineSetsToUpdate)
+	c.logger.Debugf("Deleting machineset(s) %+v", machineSetsToDelete)
+
+	for _, ms := range machineSetsToCreate {
```

Comment:
eventually I think we need to go back to the async model in the original cluster controller, but will require some thought, given that we're not receiving notifications of adds/updates/deletes of those machinesets. This looks good for now.